### PR TITLE
clientv3: add 'Sync' method

### PR DESCRIPTION
**Pull request by gyuho**
Fix https://github.com/coreos/etcd/issues/5920.

### CODE REVIEW

**Comment by gyuho** [comment]
/cc @xiang90 @heyitsanthony 

I only tested this manually, and it works.

Current integration or e2e test is not flexible enough to test (https://github.com/coreos/etcd/pull/6330#r77207347):

> This is only testing that the client is getting the member list when the important part is really client connectivity. A good test of functionality would be something like:
1) boot a cluster with one member
2) attach an autosync client
3) add two more members
4) wait for sync to grab the new members
5) stop the original member
6) use the client to access the cluster (the rpc will only go through if the sync worked)

And `SetEndpoints` is already tested. `Sync` is just a wrapper.

Any suggestion?

Thanks.

File: clientv3/balancer.go [hunk 1]
```diff
@@ -96,10 +97,17 @@ func getHost2ep(eps []string) map[string]string {
 }
 
 func (b *simpleBalancer) updateAddrs(eps []string) {
+	np := getHost2ep(eps)
```
Comment (xiang90) [inline clientv3/balancer.go]: hm... we probably need to sort np. then compare it with the current one.

File: clientv3/client.go [hunk 2]
```diff
@@ -105,6 +105,43 @@ func (c *Client) SetEndpoints(eps ...string) {
 	c.balancer.updateAddrs(eps)
 }
 
+// Sync synchronizes client's endpoint.
```
Comment (xiang90) [inline clientv3/client.go]: Sync synchronizes client's endpoints with the known endpoints from the etcd membership.

File: clientv3/client.go [hunk 3]
```diff
@@ -105,6 +105,43 @@ func (c *Client) SetEndpoints(eps ...string) {
...
+func (c *Client) Sync() error {
+	ctx, cancel := context.WithTimeout(context.Background(), c.cfg.DialTimeout)
```
Comment (xiang90) [inline clientv3/client.go]: i do not think we should use dial timeout. sync should take a context like other calls. Sync(ctx context.Context)

File: clientv3/client.go [hunk 4]
```diff
@@ -105,6 +105,43 @@ func (c *Client) SetEndpoints(eps ...string) {
...
+	mresp, err := c.MemberList(ctx)
+	cancel()
+	if err != nil {
+		return err
+	}
+	epm := make(map[string]struct{})
+	for _, m := range mresp.Members {
```
Comment (xiang90) [inline clientv3/client.go]: this is a little bit fancy. do we really need to use a map to dedup?

File: clientv3/config.go [hunk 5]
```diff
@@ -28,6 +28,9 @@ type Config struct {
 	// Endpoints is a list of URLs
 	Endpoints []string
 
+	// AutoSync is the interval to update endpoints with its latest members.
+	AutoSync time.Duration
```
Comment (xiang90) [inline clientv3/config.go]: AutoSyncInterval time.Duration

File: clientv3/config.go [hunk 6]
```diff
@@ -28,6 +28,9 @@ type Config struct {
 	// Endpoints is a list of URLs
 	Endpoints []string
 
+	// AutoSync is the interval to update endpoints with its latest members.
```
Comment (xiang90) [inline clientv3/config.go]: 0 disables auto-sync

File: clientv3/config.go [hunk 7]
```diff
@@ -46,6 +49,7 @@ type Config struct {
 
 type yamlConfig struct {
 	Endpoints             []string      `json:"endpoints"`
+	AutoSync              time.Duration `json:"auto-sync"`
```
Comment (xiang90) [inline clientv3/config.go]: AutoSyncInterval

File: clientv3/client.go [hunk 8]
```diff
@@ -105,6 +105,43 @@ func (c *Client) SetEndpoints(eps ...string) {
...
+	if c.cfg.AutoSync == time.Duration(0) {
+		return
+	}
+
+	for {
+		select {
+		case <-c.ctx.Done():
+			return
+		case <-time.After(c.cfg.AutoSync):
+			c.Sync()
```
Comment (xiang90) [inline clientv3/client.go]: check error and log it.

**Comment by xiang90** [comment]
LGTM in general

File: clientv3/balancer.go (hunk 1 above)
Comment (heyitsanthony) [inline clientv3/balancer.go]: no need to sort,
```go
match := len(np) == len(b.host2ep)
for k, v := range np {
    if b.host2ep[k] != v {
        match = false
    }
}
if match { return }
```

**Comment by gyuho** [comment]
@heyitsanthony @xiang90 All addressed. PTAL. Thanks.

File: clientv3/client.go [hunk 9]
```diff
@@ -105,6 +105,37 @@ func (c *Client) SetEndpoints(eps ...string) {
...
+		return
+	}
+
+	for {
+		select {
+		case <-c.ctx.Done():
+			return
+		case <-time.After(c.cfg.AutoSyncInterval):
+			if err := c.Sync(context.Background()); err != nil {
+				logger.Println("Sync error:", err)
```
Comment (xiang90) [inline clientv3/client.go]: Auto sync endpoints failed:, err

File: clientv3/client.go [hunk 10]
```diff
@@ -105,6 +105,37 @@ func (c *Client) SetEndpoints(eps ...string) {
...
+	if c.cfg.AutoSyncInterval == time.Duration(0) {
+		return
+	}
+
+	for {
+		select {
+		case <-c.ctx.Done():
+			return
+		case <-time.After(c.cfg.AutoSyncInterval):
+			if err := c.Sync(context.Background()); err != nil {
```
Comment (xiang90) [inline clientv3/client.go]: probably give it a timeout here. 5 seconds or so seems reasonable

File: clientv3/config.go [hunk 11]
```diff
@@ -28,6 +28,10 @@ type Config struct {
 	// Endpoints is a list of URLs
 	Endpoints []string
 
+	// AutoSyncInterval is the interval to update endpoints with its latest members.
+	// 0 disables auto-sync
```
Comment (xiang90) [inline clientv3/config.go]: 0 disables auto-sync. By default auto-sync is disabled.

**Comment by xiang90** [comment]
LGTM. defer to @heyitsanthony

File: clientv3/client.go [hunk 12]
```diff
@@ -105,6 +105,38 @@ func (c *Client) SetEndpoints(eps ...string) {
...
+	if c.cfg.AutoSyncInterval == time.Duration(0) {
+		return
+	}
+
+	for {
+		select {
+		case <-c.ctx.Done():
+			return
+		case <-time.After(c.cfg.AutoSyncInterval):
+			ctx, _ := context.WithTimeout(context.Background(), 5*time.Second)
```
Comment (heyitsanthony) [inline clientv3/client.go]: `s/context.Background()/c.ctx`?

File: clientv3/client.go [hunk 13]
```diff
@@ -105,6 +105,38 @@ func (c *Client) SetEndpoints(eps ...string) {
...
+			if err := c.Sync(ctx); err != nil {
```
Comment (heyitsanthony) [inline clientv3/client.go]: `err != nil && err != c.ctx.Err()`?

**Comment by heyitsanthony** [comment]
lgtm aside from the `c.ctx` stuff. Thanks!

File: clientv3/client.go (hunk 13 above)
Comment (gyuho) [inline clientv3/client.go]: Fixed. Merging in greens. Thanks!

**Pull request merged**